src/sentry/tasks/reports.py

```diff
@@ -917,15 +1049,110 @@ def summarize(key, points):
     }
 
 
+def build_key_errors_ctx(key_events, organization):
+    # Join with DB
+    groups = Group.objects.filter(
+        id__in=map(lambda i: i[0], key_events),
+    ).all()
+
+    group_id_to_group_history = {}
+    group_history = GroupHistory.objects.filter(
+        group__id__in=map(lambda i: i[0], key_events), organization=organization
+    ).all()
+    for g in group_history:
+        group_id_to_group_history[g.group.id] = g.get_status_display()
+
+    group_id_to_group = {}
+    for group in groups:
+        group_id_to_group[group.id] = group
+
+    return [
+        {
+            "group": group_id_to_group[e[0]],
+            "count": e[1],
+            "status": group_id_to_group_history.get(e[0], "Unresolved"),
+        }
+        for e in filter(lambda e: e[0] in group_id_to_group, key_events)
+    ]
+
+
+def build_key_transactions_ctx(key_events, organization, interval, projects):
+    project_ids = map(lambda project: project.id, projects)
+    transaction_names = map(lambda p: p[0], key_events)
+
+    start, stop = interval
+    query = Query(
+        dataset=Dataset.Transactions.value,
+        match=Entity("transactions"),
+        select=[
+            Column("transaction_name"),
+            Function("quantile(0.95)", [Column("duration")], "q95"),
+        ],
+        where=[
+            Condition(Column("finish_ts"), Op.GTE, start),
+            Condition(Column("finish_ts"), Op.LT, stop + timedelta(days=1)),
+            Condition(Column("transaction_name"), Op.IN, transaction_names),
+            Condition(Column("project_id"), Op.IN, project_ids),
+        ],
+        groupby=[Column("transaction_name")],
+    )
+    query_result = raw_snql_query(query)
+    this_week_p95 = {}
+    for point in query_result["data"]:
+        this_week_p95[point["transaction_name"]] = point["q95"]
+
+    query = Query(
+        dataset=Dataset.Transactions.value,
+        match=Entity("transactions"),
+        select=[
+            Column("transaction_name"),
+            Function("quantile(0.95)", [Column("duration")], "q95"),
+        ],
+        where=[
+            Condition(Column("finish_ts"), Op.GTE, start - timedelta(days=7)),
```

Comment:
Is there a way to merge these two queries?

Reply:
yes, you can query the events for the entire interval that you want and then split the two intervals in python

Reply:
I just need the p95 of the dataset for both intervals - how would you group by "this week" versus "last week"?

Reply:
I think this may be possible but I'll need some time to see if it is. Here is how I imagine it can be done:

Construct a query that gets all the datapoints for the two weeks, e.g.

```python3
time_range_query = Query(
    dataset=Dataset.Transactions.value,
    match=Entity("transactions"),
    select=[
        Column("transaction_name"),
        Column("timestamp"),
        Column("duration"),
    ],
    where=[
        Condition(Column("finish_ts"), Op.GTE, start - timedelta(weeks=2)),
        Condition(Column("finish_ts"), Op.LT, start),
        Condition(Column("transaction_name"), Op.IN, transaction_names),
        Condition(Column("project_id"), Op.IN, project_ids),
    ]
)
```

Use that as a subquery to query for your quantiles

```python3
quantile_queries = Query(
    dataset=Dataset.Transactions.value,
    match=time_range_query,
    ```
    
what exactly that query looks like I'll take some more time to think about.

Reply:
Ok it is possible to do this. You can change your quantile to

`quantileIf(95.0)(duration, timestamp > {something}) AS q95LastWeek` 

However it will most likely not yield a performance improvement so two queries is probably fine here. Although since they only differ by time clause I would consider refactoring so there's less duplication between the queries